In [7]:
import os
OPENAI_API_BASE= os.environ.get("OPENAI_API_BASE", "https://api-ai.numerique-interieur.com/v1") 
OPENAI_API_KEY = os.environ["OPENAI_API_KEY"]

In [10]:
from langchain_openai import ChatOpenAI
from langchain.chains.summarize import load_summarize_chain
from langchain.chat_models import ChatOpenAI
from langchain.document_loaders import WebBaseLoader
from langchain.prompts import PromptTemplate

llm = ChatOpenAI(api_key=OPENAI_API_KEY, openai_api_base=OPENAI_API_BASE, temperature=0, model="mixtral")

In [11]:
llm.invoke(
    "Why 42 ?"
)

AIMessage(content=' The number 42 is famously known as the "Answer to the Ultimate Question of Life, The Universe, and Everything" from Douglas Adams\' science fiction novel "The Hitchhiker\'s Guide to the Galaxy." In the story, a supercomputer named Deep Thought is asked to find the answer to the ultimate question of life, the universe, and everything, to which it famously replies "42," but admits that the characters aren\'t going to like it because it is very tricky and they never actually knew what the question was.\n\nIn reality, there isn\'t a deeper meaning or significance to the number 42 in this context, it is simply a humorous and nonsensical answer to a philosophical question.', response_metadata={'token_usage': {'completion_tokens': 160, 'prompt_tokens': 15, 'total_tokens': 175}, 'model_name': 'mixtral', 'system_fingerprint': None, 'finish_reason': 'stop', 'logprobs': None}, id='run-0cbcce32-b04f-428d-a06a-4c11e84584d6-0')

In [20]:
from pathlib import Path
TEXT = Path("camus.txt").read_text()
from langchain.docstore.document import Document
DOC =  Document(page_content=TEXT, metadata={"source": "local"})
DOC2 = Document(page_content=TEXT*10, metadata={"source": "local"})
llm.get_num_tokens(TEXT)

2791

## méthode langchain

In [10]:
from langchain.text_splitter import CharacterTextSplitter
from langchain.docstore.document import Document
#from langchain.llms.openai import OpenAI
from langchain.chains.summarize import load_summarize_chain

def summarize_text(text):
    # Split the text into chunks
    text_splitter = CharacterTextSplitter()
    texts = text_splitter.split_text(text)
    
    # Create document objects for each chunk
    docs = [Document(page_content=t) for t in texts]
    
    # Initialize the OpenAI model and load the summarize chain
    chain = load_summarize_chain(llm, chain_type="map_reduce")
    
    # Generate the summary
    summary = chain.run(docs)
    
    return summary


In [11]:
LLM_MAX_CONTEXT_WINDOW = 4_000

In [12]:
text = ""
num_tokens = llm.get_num_tokens(text)


SyntaxError: invalid syntax (2616875379.py, line 4)

In [13]:
from langchain.chains.combine_documents.stuff import StuffDocumentsChain
from langchain.chains.llm import LLMChain
from langchain_core.prompts import PromptTemplate

# Define prompt
prompt_template = """Write a concise summary of the following:
"{text}"
CONCISE SUMMARY:"""
prompt = PromptTemplate.from_template(prompt_template)

# Define LLM chain
llm_chain = LLMChain(llm=llm, prompt=prompt)# llm | prompt

# Define StuffDocumentsChain
stuff_chain = StuffDocumentsChain(llm_chain=llm_chain, document_variable_name="text")
stuff_chain.input_keys

['input_documents']

In [27]:
llm_chain.invoke({"text":TEXT})

KeyboardInterrupt: 

In [21]:
stuff_chain.invoke({"input_documents":[DOC2]})

{'input_documents': [Document(page_content='En recevant la distinction dont votre libre Académie a bien voulu m’honorer, ma gratitude\nétait d’autant plus profonde que je mesurais à quel point cette récompense dépassait mes\nmérites personnels. Tout homme et, à plus forte raison, tout artiste, désire être reconnu. Je\nle désire aussi. Mais il ne m’a pas été possible d’apprendre votre décision sans comparer\nson retentissement à ce que je suis réellement. Comment un homme presque jeune, riche\nde ses seuls doutes et d’une œuvre encore en chantier, habitué à vivre dans la solitude du\ntravail ou dans les retraites de l’amitié, n’aurait-il pas appris avec une sorte de panique un\narrêt qui le portait d’un coup, seul et réduit à lui-même, au centre d’une lumière crue ? De\nquel cœur aussi pouvait-il recevoir cet honneur à l’heure où, en Europe, d’autres écrivains,\nparmi les plus grands, sont réduits au silence, et dans le temps même où sa terre natale\nconnaît un malheur incessant ?\nJ’ai

In [22]:
template = """"""

prompt_template = PromptTemplate.from_template(
    """SENTENCE: {sentence}
Is the sentence true, according to the context provided above? Answer only by Yes or No, without justification.""")
chain = LLMChain(llm=llm, prompt=prompt_template)

Document(page_content='En recevant la distinction dont votre libre Académie a bien voulu m’honorer, ma gratitude\nétait d’autant plus profonde que je mesurais à quel point cette récompense dépassait mes\nmérites personnels. Tout homme et, à plus forte raison, tout artiste, désire être reconnu. Je\nle désire aussi. Mais il ne m’a pas été possible d’apprendre votre décision sans comparer\nson retentissement à ce que je suis réellement. Comment un homme presque jeune, riche\nde ses seuls doutes et d’une œuvre encore en chantier, habitué à vivre dans la solitude du\ntravail ou dans les retraites de l’amitié, n’aurait-il pas appris avec une sorte de panique un\narrêt qui le portait d’un coup, seul et réduit à lui-même, au centre d’une lumière crue ? De\nquel cœur aussi pouvait-il recevoir cet honneur à l’heure où, en Europe, d’autres écrivains,\nparmi les plus grands, sont réduits au silence, et dans le temps même où sa terre natale\nconnaît un malheur incessant ?\nJ’ai connu ce désarroi et

In [71]:
from langchain import hub
from langchain_community.document_loaders import WebBaseLoader
from langchain_chroma import Chroma
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
from langchain_openai import OpenAIEmbeddings
from langchain_text_splitters import RecursiveCharacterTextSplitter
text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
splits = text_splitter.split_documents([DOC])

import chromadb.utils.embedding_functions as embedding_functions
openai_ef = embedding_functions.OpenAIEmbeddingFunction(
                api_key="tom evil",
                api_base="https://text-embeddings.c0.cloud-pi-native.com/",
            )

In [72]:
r = openai_ef(input="i am here")

In [ ]:
#vectorstore = Chroma.from_documents(documents=splits, embedding=)
vectorstore = Chroma.from_documents(documents=splits, embedding_function=openai_ef) # https://api.python.langchain.com/en/latest/vectorstores/langchain_community.vectorstores.chroma.Chroma.html

AttributeError: 'OpenAIEmbeddingFunction' object has no attribute 'embed_documents'

In [33]:
OpenAIEmbeddings()

OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x7f17fc4ac260>, async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x7f17fc882e40>, model='text-embedding-ada-002', dimensions=None, deployment='text-embedding-ada-002', openai_api_version='', openai_api_base=None, openai_api_type='', openai_proxy='', embedding_ctx_length=8191, openai_api_key=SecretStr('**********'), openai_organization=None, allowed_special=set(), disallowed_special='all', chunk_size=1000, max_retries=2, request_timeout=None, headers=None, tiktoken_enabled=True, tiktoken_model_name=None, show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None, http_async_client=None)

In [41]:
openai_ef("troll")

AuthenticationError: Error code: 401 - {'error': {'message': "Authentication Error, API Key not allowed to access model. This token can only access models=['gpt-3.5-turbo', 'gpt-4', 'mixtral', 'vicuna', 'vicuna-33b']. Tried to access text-embedding-ada-002", 'type': 'auth_error', 'param': 'None', 'code': 401}}


https://docs.trychroma.com/embeddings/openai

Error code: 401 - {'error': {'message': "Authentication Error, API Key not allowed to access model. This token can only access models=['gpt-3.5-turbo', 'gpt-4', 'mixtral', 'vicuna', 'vicuna-33b']. Tried to access text-embedding-ada-002", 'type': 'auth_error', 'param': 'None', 'code': 401}}


In [1]:
from langchain_community.embeddings import HuggingFaceEmbeddings
embeddings = HuggingFaceEmbeddings(model_name="OrdalieTech/Solon-embeddings-base-0.1") # plus de 13 min
# tester l'API

model.safetensors:   0%|          | 0.00/1.11G [00:00<?, ?B/s]